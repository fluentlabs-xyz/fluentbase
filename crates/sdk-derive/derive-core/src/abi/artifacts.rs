use chrono::Local;
use convert_case::{Case, Casing};
use serde_json::Value;
use std::{collections::HashSet, fs, path::Path};

const INTERFACE_TEMPLATE: &str = r#"// SPDX-License-Identifier: MIT
// This file is auto-generated
// DO NOT EDIT THIS FILE MANUALLY!
// Source contract: {contract_name}
// Generated at: {timestamp}

pragma solidity ^0.8.0;

{imports}
interface I{contract_name} {
{structs}
{functions}
}
"#;

const STRUCT_TEMPLATE: &str = r#"    struct {name} {
{fields}
    }"#;

const FUNCTION_TEMPLATE: &str = r#"    function {name}({params}) external{mutability}{returns};"#;

#[derive(Debug)]
pub struct ContractArtifacts {
    pub base_name: String,
    pub abi_content: String,
    pub interface_content: String,
}

#[derive(Debug)]
struct SolidityFunction {
    name: String,
    params: Vec<SolidityParameter>,
    returns: Vec<SolidityParameter>,
    is_view: bool,
    is_payable: bool,
}

#[derive(Debug)]
struct SolidityParameter {
    name: String,
    type_name: String,
    location: Option<DataLocation>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Copy)]
enum DataLocation {
    Memory,
    Calldata,
    Storage,
}

impl ContractArtifacts {
    pub fn new(name: &str, abi: &[Value]) -> Self {
        let generator = SolidityInterfaceGenerator::new(name);
        Self {
            base_name: name.to_case(Case::Pascal),
            abi_content: serde_json::to_string_pretty(abi).expect("Failed to serialize ABI"),
            interface_content: generator.generate_from_abi(abi),
        }
    }

    pub fn save_to_dir(&self, dir: &Path) -> std::io::Result<()> {
        let contract_dir = dir.join(&self.base_name);
        fs::create_dir_all(&contract_dir)?;

        fs::write(
            contract_dir.join(format!("{}.abi.json", self.base_name)),
            &self.abi_content,
        )?;

        fs::write(
            contract_dir.join(format!("I{}.sol", self.base_name)),
            &self.interface_content,
        )?;

        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct SolidityInterfaceGenerator {
    contract_name: String,
    imports: Vec<String>,
}

impl SolidityInterfaceGenerator {
    pub fn new(name: &str) -> Self {
        Self {
            contract_name: name.to_case(Case::Pascal),
            imports: Vec::new(),
        }
    }

    pub fn generate_from_abi(&self, abi: &[Value]) -> String {
        let imports = self
            .imports
            .iter()
            .map(|import| format!("import \"{}\";", import))
            .collect::<Vec<_>>()
            .join("\n");

        let structs = self.collect_and_generate_structs(abi);
        let functions = self.generate_functions(abi);

        INTERFACE_TEMPLATE
            .replace("{contract_name}", &self.contract_name)
            .replace(
                "{timestamp}",
                &Local::now().format("%Y-%m-%d %H:%M:%S").to_string(),
            )
            .replace("{imports}", &imports)
            .replace("{structs}", &structs)
            .replace("{functions}", &functions)
    }

    fn collect_and_generate_structs(&self, abi: &[Value]) -> String {
        let struct_names = self.collect_struct_names(abi);
        let struct_definitions: Vec<String> = struct_names
            .iter()
            .filter_map(|name| self.generate_struct_definition(name))
            .collect();

        if struct_definitions.is_empty() {
            String::new()
        } else {
            format!("{}\n", struct_definitions.join("\n\n"))
        }
    }

    fn collect_struct_names(&self, abi: &[Value]) -> HashSet<String> {
        let mut seen_structs = HashSet::new();

        for func in abi {
            if let Some(inputs) = func.get("inputs").and_then(Value::as_array) {
                Self::collect_struct_types(inputs, &mut seen_structs);
            }
            if let Some(outputs) = func.get("outputs").and_then(Value::as_array) {
                Self::collect_struct_types(outputs, &mut seen_structs);
            }
        }

        seen_structs
    }

    fn collect_struct_types(params: &[Value], seen_structs: &mut HashSet<String>) {
        for param in params {
            if param["type"] == "tuple" {
                if let Some(components) = param.get("components").and_then(Value::as_array) {
                    Self::collect_struct_types(components, seen_structs);

                    if let Some(name) = param
                        .get("internalType")
                        .and_then(Value::as_str)
                        .filter(|s| s.starts_with("struct "))
                    {
                        seen_structs.insert(name[7..].to_string());
                    }
                }
            }
        }
    }

    fn generate_struct_definition(&self, struct_name: &str) -> Option<String> {
        let out_dir = std::env::var("OUT_DIR").ok()?;
        let file_path = Path::new(&out_dir)
            .join("solidity_abi")
            .join(format!("{}.json", struct_name));

        let content = fs::read_to_string(file_path).ok()?;
        let def: Value = serde_json::from_str(&content).ok()?;
        let components = def.get("components")?.as_array()?;

        let fields = components
            .iter()
            .map(|component| {
                let name = component["name"].as_str().unwrap();
                let type_str = Self::get_type_string(component);
                format!("        {} {};", type_str, name)
            })
            .collect::<Vec<_>>()
            .join("\n");

        Some(
            STRUCT_TEMPLATE
                .replace("{name}", struct_name)
                .replace("{fields}", &fields),
        )
    }

    fn generate_functions(&self, abi: &[Value]) -> String {
        abi.iter()
            .filter_map(|func| self.parse_function(func))
            .map(|func| self.format_function(&func))
            .collect::<Vec<_>>()
            .join("\n")
    }

    fn parse_function(&self, func: &Value) -> Option<SolidityFunction> {
        let name = func.get("name")?.as_str()?.to_string();
        let inputs = func.get("inputs")?.as_array()?;
        let outputs = func.get("outputs")?.as_array()?;
        let state_mutability = func.get("stateMutability").and_then(Value::as_str);

        let params = inputs
            .iter()
            .map(|param| self.parse_parameter(param))
            .collect();

        let returns = outputs
            .iter()
            .map(|param| self.parse_parameter(param))
            .collect();

        Some(SolidityFunction {
            name,
            params,
            returns,
            is_view: state_mutability == Some("view"),
            is_payable: state_mutability == Some("payable"),
        })
    }

    fn parse_parameter(&self, param: &Value) -> SolidityParameter {
        let name = param["name"].as_str().unwrap_or("").to_string();
        let type_str = Self::get_type_string(param);
        let location = self.get_data_location(param);

        SolidityParameter {
            name,
            type_name: type_str,
            location,
        }
    }

    fn format_function(&self, func: &SolidityFunction) -> String {
        let params = func
            .params
            .iter()
            .map(|param| self.format_parameter(param, true))
            .collect::<Vec<_>>()
            .join(", ");

        let mutability = if func.is_payable {
            " payable"
        } else if func.is_view {
            " view"
        } else {
            ""
        };

        let returns = if func.returns.is_empty() {
            String::new()
        } else {
            let return_params = func
                .returns
                .iter()
                .map(|param| self.format_parameter(param, false))
                .collect::<Vec<_>>()
                .join(", ");
            format!(" returns ({})", return_params)
        };

        FUNCTION_TEMPLATE
            .replace("{name}", &func.name)
            .replace("{params}", &params)
            .replace("{mutability}", mutability)
            .replace("{returns}", &returns)
    }

    fn format_parameter(&self, param: &SolidityParameter, is_input: bool) -> String {
        let location = param.location.map_or(String::new(), |loc| {
            format!(
                " {}",
                match (is_input, loc) {
                    (true, DataLocation::Memory) => "memory",
                    (true, DataLocation::Calldata) => "calldata",
                    (false, _) => "memory",
                    (_, DataLocation::Storage) => "storage",
                }
            )
        });

        if param.name.is_empty() {
            format!("{}{}", param.type_name, location)
        } else {
            format!("{}{} {}", param.type_name, location, param.name)
        }
    }

    fn get_type_string(param: &Value) -> String {
        let type_str = param["type"].as_str().unwrap();

        if type_str == "tuple" {
            if let Some(internal_type) = param
                .get("internalType")
                .and_then(Value::as_str)
                .filter(|s| s.starts_with("struct "))
            {
                return internal_type[7..].to_string();
            }

            let components = param["components"].as_array().unwrap();
            let component_types = components
                .iter()
                .map(Self::get_type_string)
                .collect::<Vec<_>>();

            format!("({})", component_types.join(","))
        } else {
            type_str.to_string()
        }
    }

    fn get_data_location(&self, param: &Value) -> Option<DataLocation> {
        let type_str = param["type"].as_str().unwrap();
        match type_str {
            "string" | "bytes" => Some(DataLocation::Calldata),
            "tuple" => Some(DataLocation::Memory),
            _ if type_str.ends_with("[]") => Some(DataLocation::Memory),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use serial_test::serial;
    use tempfile::TempDir;

    #[test]
    fn test_simple_function() {
        let abi = vec![json!({
            "name": "greeting",
            "type": "function",
            "stateMutability": "view",
            "inputs": [
                {"name": "name", "type": "string"}
            ],
            "outputs": [
                {"name": "result", "type": "string"}
            ]
        })];

        let generator = SolidityInterfaceGenerator::new("Greeter");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("interface IGreeter {"));
        assert!(interface.contains(
            "function greeting(string calldata name) external view returns (string memory result);"
        ));
    }

    #[test]
    #[serial]
    fn test_struct_generation() {
        let temp_dir = TempDir::new().unwrap();
        std::env::set_var("OUT_DIR", temp_dir.path());

        let abi_dir = temp_dir.path().join("solidity_abi");
        fs::create_dir_all(&abi_dir).unwrap();

        let point_abi = json!({
            "type": "tuple",
            "components": [
                {"name": "x", "type": "uint64"},
                {"name": "y", "type": "uint256"}
            ]
        });

        fs::write(
            abi_dir.join("Point.json"),
            serde_json::to_string_pretty(&point_abi).unwrap(),
        )
        .unwrap();

        let abi = vec![json!({
            "name": "setPoint",
            "type": "function",
            "inputs": [{
                "name": "point",
                "type": "tuple",
                "internalType": "struct Point",
                "components": [
                    {"name": "x", "type": "uint64"},
                    {"name": "y", "type": "uint256"}
                ]
            }],
            "outputs": []
        })];

        let generator = SolidityInterfaceGenerator::new("PointManager");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("struct Point {"));
        assert!(interface.contains("uint64 x;"));
        assert!(interface.contains("uint256 y;"));
        assert!(interface.contains("function setPoint(Point memory point) external;"));
    }

    #[test]
    #[serial]
    fn test_nested_structs() {
        let temp_dir = TempDir::new().unwrap();
        std::env::set_var("OUT_DIR", temp_dir.path());

        let abi_dir = temp_dir.path().join("solidity_abi");
        fs::create_dir_all(&abi_dir).unwrap();

        let point_abi = json!({
            "type": "tuple",
            "components": [
                {"name": "x", "type": "uint64"},
                {"name": "y", "type": "uint256"}
            ]
        });

        let complex_point_abi = json!({
            "type": "tuple",
            "components": [
                {
                    "name": "point",
                    "type": "tuple",
                    "internalType": "struct Point",
                    "components": [
                        {"name": "x", "type": "uint64"},
                        {"name": "y", "type": "uint256"}
                    ]
                },
                {"name": "description", "type": "string"},
                {"name": "active", "type": "bool"}
            ]
        });

        fs::write(
            abi_dir.join("Point.json"),
            serde_json::to_string_pretty(&point_abi).unwrap(),
        )
        .unwrap();

        fs::write(
            abi_dir.join("ComplexPoint.json"),
            serde_json::to_string_pretty(&complex_point_abi).unwrap(),
        )
        .unwrap();

        let abi = vec![json!({
            "name": "processPoint",
            "type": "function",
            "inputs": [{
                "name": "complexPoint",
                "type": "tuple",
                "internalType": "struct ComplexPoint",
                "components": [
                    {
                        "name": "point",
                        "type": "tuple",
                        "internalType": "struct Point",
                        "components": [
                            {"name": "x", "type": "uint64"},
                            {"name": "y", "type": "uint256"}
                        ]
                    },
                    {"name": "description", "type": "string"},
                    {"name": "active", "type": "bool"}
                ]
            }],
            "outputs": [{"name": "success", "type": "bool"}]
        })];

        let generator = SolidityInterfaceGenerator::new("PointProcessor");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("struct Point {"));
        assert!(interface.contains("struct ComplexPoint {"));

        assert!(interface.contains("uint64 x;"));
        assert!(interface.contains("uint256 y;"));
        assert!(interface.contains("Point point;"));
        assert!(interface.contains("string description;"));
        assert!(interface.contains("bool active;"));

        assert!(interface.contains("function processPoint(ComplexPoint memory complexPoint) external returns (bool success);"));
    }

    #[test]
    fn test_array_types() {
        let abi = vec![json!({
            "name": "processArrays",
            "type": "function",
            "inputs": [
                {"name": "numbers", "type": "uint256[]"},
                {"name": "names", "type": "string[]"}
            ],
            "outputs": [
                {"name": "result", "type": "bool[]"}
            ]
        })];

        let generator = SolidityInterfaceGenerator::new("ArrayProcessor");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains(
            "function processArrays(uint256[] memory numbers, string[] memory names) external returns (bool[] memory result);"
        ));
    }

    #[test]
    fn test_anonymous_tuple() {
        let abi = vec![json!({
            "name": "processTuple",
            "type": "function",
            "inputs": [{
                "name": "data",
                "type": "tuple",
                "components": [
                    {"name": "value", "type": "uint256"},
                    {"name": "flag", "type": "bool"}
                ]
            }],
            "outputs": []
        })];

        let generator = SolidityInterfaceGenerator::new("TupleProcessor");
        let interface = generator.generate_from_abi(&abi);

        assert!(interface.contains("function processTuple((uint256,bool) memory data) external;"));
    }

    #[test]
    #[serial]
    fn test_artifacts_saving() {
        let temp_dir = TempDir::new().unwrap();

        let abi = vec![json!({
            "name": "test",
            "type": "function",
            "inputs": [],
            "outputs": []
        })];

        let artifacts = ContractArtifacts::new("TestContract", &abi);
        artifacts.save_to_dir(temp_dir.path()).unwrap();

        let contract_dir = temp_dir.path().join("TestContract");
        assert!(contract_dir.exists());
        assert!(contract_dir.join("TestContract.abi.json").exists());
        assert!(contract_dir.join("ITestContract.sol").exists());
    }
}
