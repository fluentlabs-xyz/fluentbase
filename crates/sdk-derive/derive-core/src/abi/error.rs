#[derive(Debug, thiserror::Error)]
pub enum ABIError {
    #[error("Invalid type conversion: {0}")]
    TypeConversion(String),

    #[error("Serialization error: {0}")]
    Serialization(String),

    #[error("Deserialization error: {0}")]
    Deserialization(String),

    #[error("Unsupported type: {0}")]
    UnsupportedType(String),

    #[error("Unsupported pattern: {0}")]
    UnsupportedPattern(String),

    #[error("Syntax error: {0}")]
    Syntax(String),

    #[error("Internal error: {0}")]
    Internal(String),

    #[error("Config error: {0}")]
    Config(String),
}

pub type ABIResult<T> = std::result::Result<T, ABIError>;
