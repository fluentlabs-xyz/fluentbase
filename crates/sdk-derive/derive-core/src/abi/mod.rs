//! ABI module provides functionality for working with Solidity ABI, focusing on function signatures
//! parsing and type conversions.
//!
//! # Core components
//!
//! * `SolType` - represents Solidity types, parses Rust types into their Solidity equivalents
//! * `FunctionABI` - represents Solidity function definitions
//! * `RustToSol` - converts Rust types to their Solidity equivalents using registry
//!
//! # Function ID Generation
//!
//! Module can generate function signatures and IDs from Rust code in two modes:
//!
//! 1. Registry mode (enabled via feature flag):
//!    - All structure definitions are loaded into registry at startup
//!    - When constructing `FunctionABI`, registry is used to resolve types
//!    - Provides complete type information for complex structures
//!
//! 2. Direct mode (default):
//!    - No registry preloading
//!    - Types are converted directly during `FunctionABI` construction
//!    - Suitable for simple cases without complex nested structures
//!
//! ```rust, ignore
//! use crate::abi::{FunctionABI, RustToSol};
//!
//! // Parse Rust function signature
//! let sig: syn::Signature = parse_quote! {
//!     fn transfer(amount: u64, recipient: String) -> String
//! };
//!
//! // Create registry (optional)
//! // check artifacts for structure definitions
//!
//!  let config = ArtifactsRegistryConfig::new("OUT_DIR").with_mirror(".artifacts");
//! let registry = ArtifactsRegistry::new(config)?;
//!
//!
//! // Convert to FunctionABI
//! let abi = FunctionABI::from_syn(&sig, registry)?;
//!
//! // Get function ID (first 4 bytes of keccak256 hash)
//! let function_id = abi.function_id();
//!
//! // Get canonical function signature (e.g. "transfer(uint256,string)")
//! let signature = abi.signature();
//! ```
//!
//! # Constraints
//!
//! - Types must implement `SolidityABI` derive macro
//! - Generic types are not supported
//! - Module path used for type identification in registry
//! - Only non-generic Rust types can be converted to Solidity types
//! - Complex types must be registered in the registry when using Registry mode
pub mod contract;
pub mod error;
pub mod function;
pub mod parameter;
pub mod types;
