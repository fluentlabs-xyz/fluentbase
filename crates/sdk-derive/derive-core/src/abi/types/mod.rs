mod conversion;
mod sol;

// Re-export main types and conversion functions
pub use conversion::*;
pub use sol::SolType;
