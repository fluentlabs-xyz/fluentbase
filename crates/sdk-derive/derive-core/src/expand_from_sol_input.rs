use crate::abi::{
    error::ABIError,
    types::{convert_solidity_type, sol_to_rust},
};
use alloy_sol_macro_input::{SolInput, SolInputKind};
use convert_case::{Case, Casing};
use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::Ident;
use syn_solidity::{
    visit::{visit_file, Visit},
    File,
    Item,
    ItemFunction,
    Mutability,
    Spanned,
    VariableDeclaration,
};
pub struct FunctionCollector<'a> {
    pub functions: Vec<&'a ItemFunction>,
}

impl<'a> Visit<'a> for FunctionCollector<'a> {
    fn visit_item_function(&mut self, func: &'a ItemFunction) {
        self.functions.push(func);
    }
}

/// Expand the input from Solidity file or JSON ABI to a Rust trait.
pub fn expand_from_sol_input(input: SolInput) -> syn::Result<TokenStream> {
    let file = match input.kind {
        SolInputKind::Sol(sol_file) => sol_file,
        SolInputKind::Type(_) => {
            return Err(syn::Error::new(
                proc_macro2::Span::call_site(),
                "Expected Solidity interface or contract, not type",
            ));
        }
        #[cfg(feature = "json")]
        SolInputKind::Json(_, _) => {
            return Err(syn::Error::new(
                proc_macro2::Span::call_site(),
                "JSON ABI not supported in this macro",
            ));
        }
    };

    let mut visitor = FunctionCollector { functions: vec![] };
    visit_file(&mut visitor, &file);

    let mut trait_fns = Vec::new();
    for func in visitor.functions {
        let tokens = generate_trait_function(func)?;
        trait_fns.push(tokens);
    }

    let trait_name = resolve_trait_name(&file);

    Ok(quote! {
        pub trait #trait_name {
            #(#trait_fns)*
        }
    })
}

pub fn resolve_trait_name(file: &File) -> Ident {
    for item in &file.items {
        if let Item::Contract(contract) = item {
            if contract.kind.is_interface() {
                return format_ident!("{}", contract.name.to_string().to_case(Case::Pascal));
            } else if contract.kind.is_contract() {
                return format_ident!("I{}", contract.name.to_string().to_case(Case::Pascal));
            }
        }
    }

    format_ident!("GeneratedSolidityTrait")
}

pub fn resolve_receiver(func: &ItemFunction) -> TokenStream {
    if let Some(Mutability::View(_)) | Some(Mutability::Pure(_)) = func.attributes.mutability() {
        quote! { &self }
    } else {
        quote! { &mut self }
    }
}

pub fn generate_trait_function(func: &ItemFunction) -> syn::Result<TokenStream> {
    let Some(name) = &func.name else {
        return Ok(quote! {});
    };
    if name == "fallback" || name == "receive" {
        return Ok(quote! {});
    }

    let fn_name = format_ident!("{}", name.to_string().to_case(Case::Snake));
    let receiver = resolve_receiver(func);

    let mut args = Vec::new();
    for (i, param) in func.parameters.iter().enumerate() {
        args.push(generate_fn_param(i, param)?);
    }

    let ret = generate_fn_return(func)?;

    Ok(quote! {
        fn #fn_name(#receiver #(, #args)*) #ret;
    })
}

pub fn generate_fn_param(index: usize, param: &VariableDeclaration) -> syn::Result<TokenStream> {
    let name_str = param
        .name
        .as_ref()
        .map(|n| n.to_string().to_case(Case::Snake))
        .unwrap_or_else(|| format!("_param{index}"));
    let name_ident = format_ident!("{}", name_str);

    let sol_ty = convert_solidity_type(&param.ty)?;
    let rust_ty = sol_to_rust(&sol_ty).map_err(|e: ABIError| {
        syn::Error::new(param.ty.span(), format!("Cannot convert param type: {e}"))
    })?;

    Ok(quote! { #name_ident: #rust_ty })
}

pub fn generate_fn_return(func: &ItemFunction) -> syn::Result<TokenStream> {
    let Some(returns) = &func.returns else {
        return Ok(quote! {});
    };

    let return_params = &returns.returns;

    if return_params.len() == 1 {
        let sol_ty = convert_solidity_type(&return_params[0].ty)?;
        let rust_ty = sol_to_rust(&sol_ty).map_err(|e: ABIError| {
            syn::Error::new(
                return_params[0].ty.span(),
                format!("Return type error: {e}"),
            )
        })?;
        Ok(quote! { -> #rust_ty })
    } else {
        let rust_types: Vec<_> = return_params
            .iter()
            .map(|param| {
                let sol_ty = convert_solidity_type(&param.ty)?;
                sol_to_rust(&sol_ty).map(|ty| quote! { #ty }).map_err(|e| {
                    syn::Error::new(param.ty.span(), format!("Tuple return type error: {e}"))
                })
            })
            .collect::<syn::Result<_>>()?;

        Ok(quote! { -> (#(#rust_types),*) })
    }
}
