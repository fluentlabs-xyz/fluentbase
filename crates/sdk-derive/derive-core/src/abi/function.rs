use super::types::{rust_to_sol, ConversionError};
use crate::abi::{error::ABIError, parameter::Parameter};
use convert_case::{Case, Casing};
use crypto_hashes::{digest::Digest, sha3::Keccak256};
use serde::{Deserialize, Serialize};
use syn::{FnArg, Pat, ReturnType, Signature, Type};

pub const FUNCTION_ABI_TYPE: &str = "function";
/// Represents a function in the Solidity ABI
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FunctionABI {
    /// Function name in camelCase
    pub name: String,

    /// Input parameters of the function
    pub inputs: Vec<Parameter>,

    /// Output parameters of the function
    pub outputs: Vec<Parameter>,

    /// State mutability (pure, view, nonpayable, payable)
    #[serde(rename = "stateMutability")]
    pub state_mutability: StateMutability,

    /// Function type (always "function" for regular functions)
    #[serde(rename = "type")]
    pub fn_type: String,
}

/// Represents state mutability in Solidity
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum StateMutability {
    /// Can't read state
    Pure,
    /// Can read but not modify state
    View,
    /// Can modify state
    NonPayable,
    /// Can receive ETH
    Payable,
}

impl From<ConversionError> for ABIError {
    fn from(err: ConversionError) -> Self {
        ABIError::TypeConversion(err.to_string())
    }
}

impl FunctionABI {
    pub fn from_signature(sig: &Signature) -> Result<Self, ABIError> {
        Ok(Self {
            name: sig.ident.to_string().to_case(Case::Camel),
            inputs: Self::convert_inputs(&sig.inputs.iter().collect::<Vec<_>>())?,
            outputs: Self::convert_outputs(&sig.output)?,
            state_mutability: StateMutability::NonPayable,
            fn_type: FUNCTION_ABI_TYPE.to_string(),
        })
    }

    fn convert_inputs(inputs: &[&FnArg]) -> Result<Vec<Parameter>, ABIError> {
        inputs
            .iter()
            .enumerate()
            .filter_map(|(index, arg)| match arg {
                FnArg::Typed(pat_type) => {
                    let name = match &*pat_type.pat {
                        Pat::Ident(pat_ident) => pat_ident.ident.to_string(),
                        _ => format!("_{index}"),
                    };
                    Some(
                        rust_to_sol(&pat_type.ty)
                            .map(|sol_type| Parameter::new(sol_type, name))
                            .map_err(ABIError::from),
                    )
                }
                FnArg::Receiver(_) => None,
            })
            .collect()
    }

    fn convert_outputs(output: &ReturnType) -> Result<Vec<Parameter>, ABIError> {
        match output {
            ReturnType::Default => Ok(vec![]),
            ReturnType::Type(_, ty) => match &**ty {
                Type::Tuple(tuple) => tuple
                    .elems
                    .iter()
                    .enumerate()
                    .map(|(i, ty)| {
                        rust_to_sol(ty)
                            .map(|sol_type| Parameter::new(sol_type, format!("_{i}")))
                            .map_err(ABIError::from)
                    })
                    .collect(),
                _ => Ok(vec![Parameter::new(
                    rust_to_sol(ty).map_err(ABIError::from)?,
                    "_0".to_string(),
                )]),
            },
        }
    }

    /// Returns canonical function signature for Solidity ABI
    /// Format: fnName(type1,type2,...)
    pub fn signature(&self) -> Result<String, ABIError> {
        let params = self
            .inputs
            .iter()
            .map(super::parameter::Parameter::get_canonical_type)
            .collect::<Result<Vec<_>, _>>()?;

        Ok(format!("{}({})", self.name, params.join(",")))
    }

    /// Calculates function selector (first 4 bytes of keccak256 hash)
    pub fn function_id(&self) -> Result<[u8; 4], ABIError> {
        let signature = self.signature()?;
        let mut hasher = Keccak256::new();
        hasher.update(signature.as_bytes());
        let result = hasher.finalize();

        let mut selector = [0u8; 4];
        selector.copy_from_slice(&result[..4]);
        Ok(selector)
    }

    pub fn to_json(&self) -> Result<String, ABIError> {
        serde_json::to_string(self).map_err(|e| ABIError::Serialization(e.to_string()))
    }

    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }

    pub fn to_json_value(&self) -> Result<serde_json::Value, serde_json::Error> {
        serde_json::to_value(self)
    }

    pub fn from_json_value(value: serde_json::Value) -> Result<Self, serde_json::Error> {
        serde_json::from_value(value)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use syn::parse_quote;

    #[test]
    fn test_basic_function_abi() {
        let sig: Signature = parse_quote! {
            fn transfer(to: Address, amount: U256) -> bool
        };

        let abi = FunctionABI::from_signature(&sig).unwrap();

        assert_eq!(abi.name, "transfer");
        assert_eq!(abi.inputs.len(), 2);
        assert_eq!(abi.outputs.len(), 1);

        // Check inputs
        assert_eq!(abi.inputs[0].name, "to");
        assert_eq!(abi.inputs[0].ty, "address");
        assert_eq!(abi.inputs[1].name, "amount");
        assert_eq!(abi.inputs[1].ty, "uint256");

        // Check output
        assert_eq!(abi.outputs[0].name, "_0");
        assert_eq!(abi.outputs[0].ty, "bool");

        // Check signature and selector
        assert_eq!(abi.signature().unwrap(), "transfer(address,uint256)");
        assert_eq!(abi.function_id().unwrap(), [0xa9, 0x05, 0x9c, 0xbb]);
    }

    #[test]
    fn test_function_with_tuple_return() {
        let sig: Signature = parse_quote! {
            fn get_pair() -> (Address, U256)
        };

        let abi = FunctionABI::from_signature(&sig).unwrap();

        assert_eq!(abi.name, "getPair");
        assert_eq!(abi.inputs.len(), 0);
        assert_eq!(abi.outputs.len(), 2);

        // Check outputs
        assert_eq!(abi.outputs[0].name, "_0");
        assert_eq!(abi.outputs[0].ty, "address");
        assert_eq!(abi.outputs[1].name, "_1");
        assert_eq!(abi.outputs[1].ty, "uint256");
    }

    #[test]
    fn test_function_with_no_return() {
        let sig: Signature = parse_quote! {
            fn initialize(admin: Address)
        };

        let abi = FunctionABI::from_signature(&sig).unwrap();

        assert_eq!(abi.name, "initialize");
        assert_eq!(abi.inputs.len(), 1);
        assert_eq!(abi.outputs.len(), 0);

        // Check input
        assert_eq!(abi.inputs[0].name, "admin");
        assert_eq!(abi.inputs[0].ty, "address");
    }

    #[test]
    fn test_function_json_serialization() {
        let sig: Signature = parse_quote! {
            fn transfer(to: Address, amount: U256) -> bool
        };

        let abi = FunctionABI::from_signature(&sig).unwrap();
        let json = abi.to_json().unwrap();
        let deserialized = FunctionABI::from_json(&json).unwrap();

        assert_eq!(abi, deserialized);
    }
}
