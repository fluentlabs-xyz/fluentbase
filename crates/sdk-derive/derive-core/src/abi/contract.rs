use crate::abi::function::FunctionABI;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Represents a stored ABI definition with metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContractABI {
    pub id: String,
    pub functions: Vec<FunctionABI>,

    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, String>,
}

#[allow(dead_code)]
impl ContractABI {
    pub fn new(id: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            functions: Vec::new(),
            metadata: HashMap::new(),
        }
    }

    #[must_use]
    pub fn with_function(mut self, function: FunctionABI) -> Self {
        self.functions.push(function);
        self
    }

    #[must_use]
    pub fn with_functions(mut self, functions: Vec<FunctionABI>) -> Self {
        self.functions.extend(functions);
        self
    }

    pub fn with_metadata(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.metadata.insert(key.into(), value.into());
        self
    }
    #[must_use]
    pub fn path(&self) -> String {
        self.id.replace("::", "/")
    }
}
