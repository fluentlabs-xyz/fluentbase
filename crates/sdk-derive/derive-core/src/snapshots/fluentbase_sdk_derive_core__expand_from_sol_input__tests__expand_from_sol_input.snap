---
source: crates/sdk-derive/derive-core/src/expand_from_sol_input.rs
expression: formatted
---
#[derive(::fluentbase_sdk::codec::Codec, Debug, Clone, PartialEq, Eq)]
pub struct Greeting {
    pub prefix: String,
    pub name: String,
}
pub trait IMyProgramWithStruct {
    fn greeting(&self, input: Greeting) -> Greeting;
    fn custom_greeting(&self, input: Greeting) -> Greeting;
}
